aten/src/ATen/detail/VariableHooksInterface.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <ATen/Registry.h>
+#include <ATen/Error.h>
+#include <ATen/ScalarType.h>
+
+namespace at {
+  class Context;
+}
+
+// NB: Registry class not actually in the namespace detail, due to limitations
+// of Registry.h
+namespace at {
+
+// The VariableHooksInterface is an interface for autograd functionality
+// which currently doesn't live in libATen.so AND needs to be called from
+// ATen.  In this case, it is only the type registry for Variable types,
```

Comment:
Why do we use VariableType anywhere in ATen?

Reply:
Yeah, it's used to add Variable variants of types whenever we register new types to the type registry.  The specific case we have to worry about looks like this:

1. libATen.so is loaded, and it initializes the type registry with CPU types.  However, CUDA types are not initialized (they are only initialized when we lazily initialize CUDA)
2. libtorch.so is loaded, and it initializes the variable type registry with all the types in the type registry, which in this case is only the CPU types
3. During execution, CUDA initialization occurs.  CUDA initializes the type registry with CUDA types, and now we need to also, at the same time we're initializing the CUDA types, also initialize the CUDA variable types.

Also, because C10 is going to have native knowledge about Variables, I sucked the variable type registry into ATen.